### ISSUE TITLE
Loading multiple models?

### ISSUE THREAD

**Issue by lumost** [guillaume-be/rust-bert#258]
The following rust snippet will result in a panic when both GPT2 and BERT are used due to an error that vs.load cannot find a tensor, the specific tensor is none deterministic.  Is there a canonical method for loading multiple models from disk?

### Error
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: TensorNameNotFound("bert.encoder.layer.2.intermediate.dense.bias", "/home/lumost/.cache/.rustbert/gpt2/model/42eae7d865748b39035c10e9e35f486abe3f4c59a139a5650b689fcbfa7fce6a.a33043547ef61535ac82c32c71f646d47516ffbb0cf52cdfe0ea174c0332582b")',
```

### Reproduction
```    
    let mut vs = VarStore::new(Device::cuda_if_available());
    
    // load bert
    let config_resource = RemoteResource::from_pretrained(BertConfigResources::BERT);
    let vocab_resource = RemoteResource::from_pretrained(BertVocabResources::BERT);
    let weights_resource = RemoteResource::from_pretrained(BertModelResources::BERT);
    let config_path = config_resource.get_local_path().unwrap();
    let vocab_path = vocab_resource.get_local_path().unwrap();
    let weights_path = weights_resource.get_local_path().unwrap();
    let config = BertConfig::from_file(config_path);
    let bert_model: BertModel<BertEmbeddings> = BertModel::new(&vs.root() / "bert", &config);
    vs.load(weights_path).unwrap();

    // load gpt2
    let config_resource = RemoteResource::from_pretrained(Gpt2ConfigResources::GPT2);
    let weights_resource = RemoteResource::from_pretrained(Gpt2ModelResources::GPT2);
    let config_path = config_resource.get_local_path().unwrap();
    let weights_path = weights_resource.get_local_path().unwrap();
    let config = Gpt2Config::from_file(config_path);
    let gpt2_model = GPT2LMHeadModel::new(&vs.root() / "gpt2", &config);
    vs.load(weights_path).unwrap();

**Comment by guillaume-be** [comment]
Hello @lumost ,

The easiest way to do so (if you are not training the model and do not propagate gradients through both models) would be to use a separate `VarStore` per model:

```rust
   
    // load bert
    let mut bert_vs = VarStore::new(Device::cuda_if_available());

    let config_resource = RemoteResource::from_pretrained(BertConfigResources::BERT);
    let vocab_resource = RemoteResource::from_pretrained(BertVocabResources::BERT);
    let weights_resource = RemoteResource::from_pretrained(BertModelResources::BERT);
    let config_path = config_resource.get_local_path().unwrap();
    let vocab_path = vocab_resource.get_local_path().unwrap();
    let weights_path = weights_resource.get_local_path().unwrap();
    let config = BertConfig::from_file(config_path);
    let bert_model: BertModel<BertEmbeddings> = BertModel::new(&vs.root() / "bert", &config);
    bert_vs.load(weights_path).unwrap();

    // load gpt2
    let mut gpt2_vs = VarStore::new(Device::cuda_if_available());
    let config_resource = RemoteResource::from_pretrained(Gpt2ConfigResources::GPT2);
    let weights_resource = RemoteResource::from_pretrained(Gpt2ModelResources::GPT2);
    let config_path = config_resource.get_local_path().unwrap();
    let weights_path = weights_resource.get_local_path().unwrap();
    let config = Gpt2Config::from_file(config_path);
    let gpt2_model = GPT2LMHeadModel::new(&vs.root() / "gpt2", &config);
    gpt2_vs.load(weights_path).unwrap();
```

Could you please give this a try and let me know if it solves the issue?

**Comment by lumost** [comment]
Hey @guillaume-be 

Unfortunately I do need to propagate the gradients through both models.  Is there a good way to load both in this context? Also happy to take a look myself if you have a good code pointer to where the problem lies.

**Comment by lumost** [comment]
Thanks for the CR @guillaume-be! One item I noticed while testing it out is that the classic Bert model has it's own loading problems.  The below code fails with a panic.

```
2022-06-08 22:23:43,928 INFO [cached_path::cache] Cached version of https://huggingface.co/bert-base-uncased/resolve/main/rust_model.ot is up-to-date
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: TensorNameNotFound("encoder.layer.11.intermediate.dense.weight", "/home/lumost/.cache/.rustbert/bert/model/3df4a644e041ac387ba497bcc7b7192c806fe76484b0847d08f55f1397d2119e.567e5c9278e5cd29212929e359997c1649290a20e4e3d7b19ffc9838e9f631a6")', src/main.rs:155:28
```


```
    let mut vs5 = VarStore::new(Device::Cpu);

    let config_resource = RemoteResource::from_pretrained(BertConfigResources::BERT);
    let vocab_resource = RemoteResource::from_pretrained(BertVocabResources::BERT);
    let weights_resource = RemoteResource::from_pretrained(BertModelResources::BERT);
    let config_path = config_resource.get_local_path().unwrap();
    let vocab_path = vocab_resource.get_local_path().unwrap();
    let weights_path = weights_resource.get_local_path().unwrap();
    let config = BertConfig::from_file(config_path);
    let bert_model: BertModel<BertEmbeddings> = BertModel::new(&vs5.root(), &config);
    vs5.load(weights_path).unwrap();
    return;
```

**Comment by guillaume-be** [comment]
Hello @lumost ,

This is because the checkpoint used for the BERT model is actually the BERT for masked language model (https://github.com/huggingface/transformers/blob/224bde91caff4ccfd12277ab5e9bf97c61e22ee9/src/transformers/models/bert/modeling_bert.py#L1292). This model stores BERT in `self.bert` and therefore adds this to the prefix of variables.

You can load instead a `BertForMaskedLM` to keep the same architecture, or may load the model by adding the `bert` prefix to the `VarStore`. The example below should wokr loading both `VarStore` and merging them into one, waiting for https://github.com/LaurentMazare/tch-rs/issues/494 to be released.

```rust
use rust_bert::bert::{
    BertConfig, BertConfigResources, BertEmbeddings, BertModel, BertModelResources,
};
use rust_bert::gpt2::{GPT2LMHeadModel, Gpt2Config, Gpt2ConfigResources, Gpt2ModelResources};
use rust_bert::resources::{RemoteResource, ResourceProvider};
use rust_bert::Config;
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use tch::nn::{VarStore, Variables};
use tch::Device;

fn main() {
    // load bert
    let mut bert_vs = VarStore::new(Device::Cpu);

    let config_resource = RemoteResource::from_pretrained(BertConfigResources::BERT);
    let weights_resource = RemoteResource::from_pretrained(BertModelResources::BERT);
    let config_path = config_resource.get_local_path().unwrap();
    let weights_path = weights_resource.get_local_path().unwrap();
    let config = BertConfig::from_file(config_path);
    let _bert_model: BertModel<BertEmbeddings> = BertModel::new(&bert_vs.root() / "bert", &config);
    bert_vs.load(weights_path).unwrap();

    // load gpt2
    let mut gpt2_vs = VarStore::new(Device::cuda_if_available());
    let config_resource = RemoteResource::from_pretrained(Gpt2ConfigResources::DISTIL_GPT2);
    let weights_resource = RemoteResource::from_pretrained(Gpt2ModelResources::DISTIL_GPT2);
    let config_path = config_resource.get_local_path().unwrap();
    let weights_path = weights_resource.get_local_path().unwrap();
    let config = Gpt2Config::from_file(config_path);
    let _gpt2_model = GPT2LMHeadModel::new(&gpt2_vs.root(), &config);
    gpt2_vs.load(weights_path).unwrap();

    // Merge both models' varstores
    
    let mut new_var_store = VarStore::new(Device::cuda_if_available());
    let mut variables = Variables {
        named_variables: HashMap::new(),
        trainable_variables: Vec::new(),
    };
    for (var_name, var) in bert_vs.variables() {
        let new_var_name = format!("bert_{}", var_name);
        variables.named_variables.insert(new_var_name, var);
    }
    for trainable_var in bert_vs
        .variables_
        .lock()
        .unwrap()
        .trainable_variables
        .drain(..)
    {
        variables.trainable_variables.push(trainable_var);
    }

    for (var_name, var) in gpt2_vs.variables() {
        let new_var_name = format!("gpt2_{}", var_name);
        variables.named_variables.insert(new_var_name, var);
    }
    for trainable_var in gpt2_vs
        .variables_
        .lock()
        .unwrap()
        .trainable_variables
        .drain(..)
    {
        variables.trainable_variables.push(trainable_var);
    }

    new_var_store.variables_ = Arc::new(Mutex::new(variables));
    println!("{:?}", new_var_store);
}
```

**Comment by guillaume-be** [comment]
Fixed with 0.19.0 release
